include/mumble/Base64.hpp

```diff
@@ -3,30 +3,134 @@
 // that can be found in the LICENSE file at the root of the
 // Mumble source tree or at <https://www.mumble.info/LICENSE>.
 
+// Inspired by https://renenyffenegger.ch/notes/development/Base64/Encoding-and-decoding-base-64-with-cpp.
+
 #ifndef MUMBLE_BASE64_HPP
 #define MUMBLE_BASE64_HPP
 
-#include "Macros.hpp"
-#include "NonCopyable.hpp"
 #include "Types.hpp"
 
-#include <memory>
+#include <array>
+#include <cmath>
+#include <limits>
+
+#include <gsl/span>
 
 namespace mumble {
-class MUMBLE_EXPORT Base64 : NonCopyable {
+class Base64 {
 public:
-	class P;
+	using StrView      = gsl::span< char >;
+	using StrViewConst = gsl::span< const char >;
+
+	static constexpr size_t decode(const BufView out, StrViewConst in) {
+		constexpr std::array< uint8_t, std::numeric_limits< uint8_t >::max() + 1 > table = {
+			64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+			64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 62, 64, 64, 64, 63, 52, 53, 54, 55,
+			56, 57, 58, 59, 60, 61, 64, 64, 64, 64, 64, 64, 64, 0,  1,  2,  3,  4,  5,  6,  7,  8,  9,  10, 11, 12,
+			13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 64, 64, 64, 64, 64, 64, 26, 27, 28, 29, 30, 31, 32,
+			33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 64, 64, 64, 64, 64, 64, 64,
+			64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+			64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+			64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+			64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+			64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64
+		};
+
+		if (in.empty() || in.size() % 4) {
+			return 0;
+		}
+
+		size_t size = in.size() / 4 * 3;
+		if (auto iter = in.rbegin(); *iter == PAD_CHAR) {
+			size -= *++iter == PAD_CHAR ? 2 : 1;
+		}
+
+		if (!out.data()) {
+			return size;
+		}
+
+		if (out.size() < size) {
+			return 0;
+		}
+
+		ByteView data(reinterpret_cast< uint8_t * >(out.data()), out.size());
```

Comment:
Why the change of type? `std::byte` seems like the more appropriate type anyway.

Reply:
Operations on `std::byte` are not allowed.

Reply:
What operations?
To be honest, the more I use `std::byte`, the less I like it. Maybe we would be better off by simply using `byte` as an alias for `uint8_t`?

Reply:
Arithmetic operations.

`std::byte` is meant to represent data that is in an unknown/unspecified format, which is actually the case here.

However, as we know what we need to do with the data, we cast to the type that is more comfy for us to work with.

Reply:
Bit operations in bytes are quite common though, so it is silly that this is not allowed 👀

Reply:
Well, that's why [`std::to_integer()`](https://en.cppreference.com/w/cpp/types/byte#std::to_integer) exists.

The main reason would be to avoid dangerous implicit casts.

Reply:
Ah, but according to [that site](https://en.cppreference.com/w/cpp/types/byte) the bitwise operators are overloaded for `std::byte`, so we shouldn't have issues in the first place? Can't we replace the addition by a binary AND?

Reply:
Yup.

Reply:
Better idea:

```c++
static constexpr std::byte operator+(const std::byte lhs, const std::byte rhs) {
	const auto ret = std::to_integer< uint8_t >(lhs) + std::to_integer< uint8_t >(rhs);
	return std::byte(ret);
}
```

Reply:
That'd also solve the problem, but imo it'd be better to use a binary AND, if that is what is being performed. A binary AND and an addition are semantically very different things, so I think one should make it explicit what is going on :thinking:

Reply:
It's really an addition, a binary AND would give you a different result.

Reply:
Ho - indeed. Yeah then using the overload for `operator+` seems good to me :+1: